### DISCUSSION TITLE
Aiohttp update

### DISCUSSION THREAD

**Discussion by FlutterGenerator** [aio-libs/aiohttp#9871] [Q&A]
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/data/data/com.termux/files/home/friendly-telegram/friendly-telegram/__main__.py", line 33, in <module>                                             from . import main                                                        File "/data/data/com.termux/files/home/friendly-telegram/friendly-telegram/main.py", line 32, in <module>
    from telethon import TelegramClient, events                               File "/data/data/com.termux/files/home/.local/lib/python3.12/site-packages/telethon/__init__.py", line 1, in <module>
    from .client.telegramclient import TelegramClient
  File "/data/data/com.termux/files/home/.local/lib/python3.12/site-packages/telethon/client/__init__.py", line 21, in <module>                             from .downloads import DownloadMethods
  File "/data/data/com.termux/files/home/.local/lib/python3.12/site-packages/telethon/client/downloads.py", line 16, in <module>
    import aiohttp                                                            File "/data/data/com.termux/files/usr/lib/python3.12/site-packages/aiohttp/__init__.py", line 6, in <module>
    from .client import (
  File "/data/data/com.termux/files/usr/lib/python3.12/site-packages/aiohttp/client.py", line 39, in <module>
    from . import hdrs, http, payload
  File "/data/data/com.termux/files/usr/lib/python3.12/site-packages/aiohttp/http.py", line 7, in <module>
    from .http_parser import (
  File "/data/data/com.termux/files/usr/lib/python3.12/site-packages/aiohttp/http_parser.py", line 49, in <module>
    from .http_writer import HttpVersion, HttpVersion10
  File "/data/data/com.termux/files/usr/lib/python3.12/site-packages/aiohttp/http_writer.py", line 230, in <module>
    _c_serialize_headers = _http_writer._serialize_headers
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'aiohttp._http_writer' has no attribute '_serialize_headers'

**Comment by webknjaz** [comment]
What's the aiohttp version and the rest of the context? I can only guess this is running the pure-python variant under Android but that's about it...

**Reply by webknjaz** [reply]
Although, the exception indicates that the C-extension ended up being loaded. I don't know why, though.

**Reply by FlutterGenerator** [reply]
ok
спасибо посмотрю

**Reply by FlutterGenerator** [reply]
|                                                      ^
      aiohttp/_websocket.c:2696:56: warning: 'ma_version_tag' is deprecated [-Wdeprecated-declarations]
       2696 |     if (unlikely(!dict) || unlikely(tp_dict_version != __PYX_GET_DICT_VERSION(dict)))
            |                                                        ^
      aiohttp/_websocket.c:1118:65: note: expanded from macro '__PYX_GET_DICT_VERSION'
       1118 | #define __PYX_GET_DICT_VERSION(dict)  (((PyDictObject*)(dict))->ma_version_tag)
            |                                                                 ^
      /data/data/com.termux/files/usr/include/python3.12/cpython/dictobject.h:22:5: note: 'ma_version_tag' has been explicitly marked deprecated here
         22 |     Py_DEPRECATED(3.12) uint64_t ma_version_tag;
            |     ^
      /data/data/com.termux/files/usr/include/python3.12/pyport.h:317:54: note: expanded from macro 'Py_DEPRECATED'
        317 | #define Py_DEPRECATED(VERSION_UNUSED) __attribute__((__deprecated__))
            |                                                      ^
      aiohttp/_websocket.c:2741:9: warning: 'ma_version_tag' is deprecated [-Wdeprecated-declarations]
       2741 |         __PYX_PY_DICT_LOOKUP_IF_MODIFIED(
            |         ^
      aiohttp/_websocket.c:1125:16: note: expanded from macro '__PYX_PY_DICT_LOOKUP_IF_MODIFIED'
       1125 |     if (likely(__PYX_GET_DICT_VERSION(DICT) == __pyx_dict_version)) {\
            |                ^
      aiohttp/_websocket.c:1118:65: note: expanded from macro '__PYX_GET_DICT_VERSION'
       1118 | #define __PYX_GET_DICT_VERSION(dict)  (((PyDictObject*)(dict))->ma_version_tag)
            |                                                                 ^
      /data/data/com.termux/files/usr/include/python3.12/cpython/dictobject.h:22:5: note: 'ma_version_tag' has been explicitly marked deprecated here
         22 |     Py_DEPRECATED(3.12) uint64_t ma_version_tag;
            |     ^
      /data/data/com.termux/files/usr/include/python3.12/pyport.h:317:54: note: expanded from macro 'Py_DEPRECATED'
        317 | #define Py_DEPRECATED(VERSION_UNUSED) __attribute__((__deprecated__))
            |                                                      ^
      aiohttp/_websocket.c:2741:9: warning: 'ma_version_tag' is deprecated [-Wdeprecated-declarations]
       2741 |         __PYX_PY_DICT_LOOKUP_IF_MODIFIED(
            |         ^
      aiohttp/_websocket.c:1129:30: note: expanded from macro '__PYX_PY_DICT_LOOKUP_IF_MODIFIED'
       1129 |         __pyx_dict_version = __PYX_GET_DICT_VERSION(DICT);\
            |                              ^
      aiohttp/_websocket.c:1118:65: note: expanded from macro '__PYX_GET_DICT_VERSION'
       1118 | #define __PYX_GET_DICT_VERSION(dict)  (((PyDictObject*)(dict))->ma_version_tag)
            |                                                                 ^
      /data/data/com.termux/files/usr/include/python3.12/cpython/dictobject.h:22:5: note: 'ma_version_tag' has been explicitly marked deprecated here
         22 |     Py_DEPRECATED(3.12) uint64_t ma_version_tag;
            |     ^
      /data/data/com.termux/files/usr/include/python3.12/pyport.h:317:54: note: expanded from macro 'Py_DEPRECATED'
        317 | #define Py_DEPRECATED(VERSION_UNUSED) __attribute__((__deprecated__))
            |                                                      ^
      aiohttp/_websocket.c:3042:55: error: no member named 'ob_digit' in 'struct _longobject'
       3042 |             const digit* digits = ((PyLongObject*)x)->ob_digit;
            |                                   ~~~~~~~~~~~~~~~~~~  ^
      aiohttp/_websocket.c:3097:55: error: no member named 'ob_digit' in 'struct _longobject'
       3097 |             const digit* digits = ((PyLongObject*)x)->ob_digit;
            |                                   ~~~~~~~~~~~~~~~~~~  ^
      aiohttp/_websocket.c:3238:55: error: no member named 'ob_digit' in 'struct _longobject'
       3238 |             const digit* digits = ((PyLongObject*)x)->ob_digit;
            |                                   ~~~~~~~~~~~~~~~~~~  ^
      aiohttp/_websocket.c:3293:55: error: no member named 'ob_digit' in 'struct _longobject'
       3293 |             const digit* digits = ((PyLongObject*)x)->ob_digit;
            |                                   ~~~~~~~~~~~~~~~~~~  ^
      aiohttp/_websocket.c:3744:47: error: no member named 'ob_digit' in 'struct _longobject'
       3744 |     const digit* digits = ((PyLongObject*)b)->ob_digit;
            |                           ~~~~~~~~~~~~~~~~~~  ^
      6 warnings and 5 errors generated.
      error: command '/data/data/com.termux/files/usr/bin/aarch64-linux-android-clang' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for aiohttp
Failed to build aiohttp
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (aiohttp) 


Eroor брат

**Reply by bdraco** [reply]
Which version of cython are you using to do the build?

**Reply by bdraco** [reply]
It looks like we don't set a minimum Cython version

We should probably pin that in the `build-system` to what we use internally so we know we get the same result for manual builds.... the downside is it will install Cython to build even if extensions are disabled and won't use it (which is probably ok)

```diff
diff --git a/pyproject.toml b/pyproject.toml
index 339626869..e8eb56dc2 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,6 +1,7 @@
 [build-system]
 requires = [
     "setuptools >= 46.4.0",
+    "Cython >= 3.0.11",
 ]
 build-backend = "setuptools.build_meta"
 ```

**Reply by bdraco** [reply]
Assuming thats the problem, I opened https://github.com/aio-libs/aiohttp/pull/9876 for that, but I don't want to move it forward until we confirm

**Reply by asvetlov** [reply]
How did Cython appear in the system at all, despite it was brand new or outdated?

@FlutterGenerator how did you install aiohttp in your system?

**Reply by FlutterGenerator** [reply]
termux mobile https://github.com/Netuzb/UMod Throws an error

**Reply by asvetlov** [reply]
Looks like @bdraco is right, old Cython version installed in the @FlutterGenerator box.
Installation on mobile is an edge case for aiohttp, we have no consensus on how to manage the problem once for everybody yet.

If you could install modern Cython it will fix the error.
Sorry, I never executed python on mobile and cannot provide comprehensive instructions.
